### PR TITLE
Add on_error option for Cholesky

### PR DESCRIPTION
**Pull request by aseyboldt**
The cholesky Op always raises an exception if the matrix is not positive definite. In some cases it can help if it returned nans instead, so that that case can be handled within the theano graph.
This was first implemented in pymc-devs/pymc3#2028
CC @nouiz

### CODE REVIEW

File: theano/tensor/slinalg.py [hunk 1]
```diff
@@ -79,6 +97,13 @@ def grad(self, inputs, gradients):
         dz = gradients[0]
         chol_x = self(x)
 
+        # Replace the cholesky decomposition with 1 if there are nans
+        # or solve_upper_triangular will throw a ValueError.
+        if self.on_error == 'nan':
+            ok = tensor.all(tensor.nlinalg.diag(chol_x) > 0)
```
Comment (nouiz) [inline theano/tensor/slinalg.py]: ok = tensor.isnan(tensor.nlinalg.diag(cho_x))

why why check only the diag? the perform set the full output to nan.

File: theano/tensor/slinalg.py [hunk 2]
```diff
@@ -97,9 +122,14 @@ def conjugate_solve_triangular(outer, inner):
             chol_x, tril_and_halve_diagonal(chol_x.T.dot(dz)))
 
         if self.lower:
-            return [tensor.tril(s + s.T) - tensor.diag(tensor.diagonal(s))]
+            grad = tensor.tril(s + s.T) - tensor.diag(tensor.diagonal(s))
+        else:
+            grad = tensor.triu(s + s.T) - tensor.diag(tensor.diagonal(s))
+
+        if self.on_error == 'nan':
+            return [tensor.switch(ok, grad, np.nan)]
```
Comment (nouiz) [inline theano/tensor/slinalg.py]: if ok is false, this won't return data in the right shape. I think you want to replace np.nan to:

theano.tensor.fill(grad, np.nan)

Comment (lamblin) [inline theano/tensor/slinalg.py]: `tensor.switch` is an Elemwise, so it will broadcast its 3 inputs to the same shape. So I think this code is correct.

Comment (nouiz) [inline theano/tensor/slinalg.py]: right

File: theano/tensor/slinalg.py (hunk 1 above)
Comment (aseyboldt) [inline theano/tensor/slinalg.py]: done

**Pull request merged**